test/emqx_auth_username_SUITE.erl

```diff
@@ -42,11 +48,243 @@ end_per_suite(_Config) ->
+                                                                    117,115,101,114,110,97,109,101,     % <<"username">>
+                                                                    0,                                  % U+0000
+                                                                    112,97,115,115,119,111,114,100>>}, undefined),
+    {error, bad_authentication_method} = emqx_access_control:authenticate(User2#{auth_method  := <<"SCRAM-SHA-1">>}, undefined),
+    {error, invalid_auth_data} = emqx_access_control:authenticate(#{auth_method => <<"PLAIN">>, auth_data => <<>>}, undefined),
+    {error, invalid_auth_data} = emqx_access_control:authenticate(#{auth_method => <<"PLAIN">>}, undefined),
     ok = emqx_auth_username:remove_user(<<"test_username">>),
-    {error, _} = emqx_access_control:authenticate(User1, <<"password">>).
+    {error, _} = emqx_access_control:authenticate(User1, <<"password">>),
+    {error, _} = emqx_access_control:authenticate(User2, <<"password">>).
+
+t_enhanced_auth(_Config) ->
+    ok = emqx_auth_username:add_user(<<"admin">>, <<"public">>),
+    % authenticate with unsupported method. will failed
+    {ok, Sock} = emqx_client_sock:connect({127,0,0,1}, 1883, [binary, {packet, raw}, {active, false}], 3000),
+    Packet = raw_send_serialise(?CONNECT_PACKET(#mqtt_packet_connect{
+                                        client_id   = <<"mqtt_client">>,
+                                        proto_ver   = ?MQTT_PROTO_V5,
+                                        clean_start = false,
+                                        properties  = #{'Authentication-Method' => <<"SCRAM-SHA-1">>,
+                                                        'Authentication-Data'   => <<0,97,100,109,105,110,          % admin
+                                                                                     0,112,117,98,108,105,99>>}})), % public
+    emqx_client_sock:send(Sock, Packet),
+    {ok, Data} = gen_tcp:recv(Sock, 0),
+    {ok, ?CONNACK_PACKET(?RC_BAD_AUTHENTICATION_METHOD), _} = raw_recv_pase(Data),
+    emqx_client_sock:close(Sock),
+
+    % authenticate with empty method. will failed
+    {ok, Sock1} = emqx_client_sock:connect({127,0,0,1}, 1883, [binary, {packet, raw}, {active, false}], 3000),
+    Packet1 = raw_send_serialise(?CONNECT_PACKET(#mqtt_packet_connect{
+                                        client_id   = <<"mqtt_client">>,
+                                        proto_ver   = ?MQTT_PROTO_V5,
+                                        clean_start = false,
+                                        properties  = #{'Authentication-Method' => <<>>,
+                                                        'Authentication-Data'   => <<>>}})),
+    emqx_client_sock:send(Sock1, Packet1),
+    {ok, Data1} = gen_tcp:recv(Sock1, 0),
+    {ok, ?CONNACK_PACKET(?RC_MALFORMED_PACKET), _} = raw_recv_pase(Data1),
+    emqx_client_sock:close(Sock1),
+
```

Comment:
There are too many duplicated procedure in your test cases, can you abstract your code and make it shorter and more clear ?

Reply:
You can just write like this
```erlang
with_connection(fun(Sock) ->

                   %% your procedure
                end).

```

Reply:
It's difficult, because it's flexible.